### PR TITLE
[Week5] Finish hw1 and HTTPGames

### PR DESCRIPTION
**Pull request by ishin4554**
**更新**
- 完成 HTTPGames 
- 完成 hw1 心得

**心得**
- **超超超超超級喜歡 HTTPGames 的啦** ，通宵一晚上破卡讓人回憶起玩掌機破 RPG 的時光，捨不得睡覺耶 ˊˇˋ 但是 code 有一點通關後的狼藉感，之後學到新東西跟著後面的關卡一起優化。
- Base64 跟中文的 Decode, Encode 是繼 BFS 之後沒有課程努力查資料的兩個小議題，很有成就感。
- 因為進度有點散佈在不同週，今天的 PR 比較瑣碎，麻煩老師了 !

### CODE REVIEW

File: homeworks/week5/httpGame/lv10.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    (err, res, body) => {
+      if (body.length !== 4) {
+        console.log(num + body);
+      }
+    });
+}
+
+
+// upload number to guess
+function guess(num, i) {
```
Comment (aszx87410) [inline homeworks/week5/httpGame/lv10.js:40]: 我其實沒有預期有人這題用程式解XDD 我把這關當遊戲在玩來解哈哈

File: homeworks/week5/hw1.md [hunk 2]
```diff
@@ -1,3 +1,381 @@
...
+#### HTTPGames
+真的很喜歡這種關卡系的設計，太好玩了，甚至應該說這種 RPG 關卡很剋我啊 XDDD 通宵肝玩了十關，但因為是一種破遊戲的心情，code 寫超醜的啦 XDDD 可是很盡興，跟以前打遊戲熬夜破卡的心情一模一樣，超級停不下來。
+
+- 一開始有一點卡關，因為有一些題目第一次閱讀上語意會不清楚：例如：第二關要找書，一開始想說是用肉眼確認嗎？所以就先用肉眼解進入下一關（後來才發現應該要用程式找比較合理），例如：base 64 範例跟實際轉換結果不同，有一度還以為是不是我的寫法錯了，卡關了一陣子，還好只是 form 忘記改成 headers
+
+> 但我想說的是，其實沒關係啦，遊戲就是要這樣才好玩啊！寫得太清楚就無趣了！
+
+- 有一些議題解決得很過癮，算是計劃中第一次在沒有課程輔助的情況下解決問題：像是中文解碼問題跟 base 64，由於不是在瀏覽器上，所以是用 Buffer 來轉換，中文解碼則看別人怎麼做換算實際寫一次。
+
+- 中間如果可以弄懂 Callback 更好的寫法就好了：常常開了變數可是 request 的東西還沒回來QQ 後面幾關的 callback 寫得超醜。有一點不滿意，但 promise 等等感覺是很高階的東西，好像不應該在這階段處理。
```
Comment (aszx87410) [inline homeworks/week5/hw1.md:287]: 可以稍微看一下 promise，大概理解一下用法之類的XD

File: homeworks/week5/hw1.md [hunk 3]
```diff
@@ -1,3 +1,381 @@
...
+
+- 好不容易最後一題的時候，居然要寫排列演算法:就好像以為苦盡甘來就要最後一關，結果最後一關是你最不擅長被剋的關卡。但解完滿痛快的，本來以為自己會不會只要是這種嚴謹的算法題目都會做不出來，幸好還解得出來 QQ
```
Comment (aszx87410) [inline homeworks/week5/hw1.md:289]: 我原本預期是當遊戲來玩 1A2B，不是程式解XDD 會容易滿多

File: homeworks/week5/hw1.md [hunk 4]
```diff
@@ -1,3 +1,381 @@
...
+
+- 好不容易最後一題的時候，居然要寫排列演算法:就好像以為苦盡甘來就要最後一關，結果最後一關是你最不擅長被剋的關卡。但解完滿痛快的，本來以為自己會不會只要是這種嚴謹的算法題目都會做不出來，幸好還解得出來 QQ 
+
+之後再找一個週末解開 true end 把第十關之後解一解。
+
+除了 HTTPGames 之外複習週主要像是總確認，確定一些學習基底的部分有沒有歪掉。
+
+#### LidemyOJ 
+
+OJ 的特色是不會回傳結果參考，所以要自行想 test case 來測試，有時候有部分分數沒有的時候最艱難，要想像自己的可能錯了什麼，尋找 edge case 有時候比改 code 還難。也有發生自己的 test 沒過但是 OJ 過了的狀況 XD
```
Comment (aszx87410) [inline homeworks/week5/hw1.md:297]: 這其實是大部分 OJ 都是這樣，不會跟你說你錯哪裡，雖然覺得很靠北但也沒辦法。畢竟現實生活中也是這樣，你不一定保證你能知道使用者輸入了什麼。

Comment (aszx87410) [reply homeworks/week5/hw1.md:297]: 然後自己 test 沒過 OJ 過了，可以把 test case 給我嗎XDD 代表我 test case 出太爛哈哈

File: homeworks/week5/hw1.md [hunk 5]
```diff
@@ -1,3 +1,381 @@
...
+
+- 服科所資訊讀書會：大概了解到工程師會關注什麼議題，例如： yarn 跟 npm，從我的角度會覺得只要知道哪個我喜歡用、有沒有什麼負面效果就好，但純種工程師會深究底層的資料結構怎麼實踐才解決了原本 npm 版本混亂的問題。其他還有聽到的內容有：Chrome 渲染、Virtual Dom、Docker 部署、JS Stream 目前只能聽懂 3, 4 成。不過 Chrome 渲染還滿有幫助的，比較知道在寫效果的時候，哪些效果可以比較快出現在螢幕上。
+
+- 第三期實體聚會：開啟了更多我對工程師的想像，我覺得這種多看看不同類型的前端工程師，對於自己釐清自己的定位很有幫助，希望可以找到靠近設計一點點的前端工程師。
+
+- 前輩線上小聊： 開始 Code Review 這個想法是來自於第二期 A 學姊跟實體聚會的建議，所以要求自己至少把值得參考看完，了解其他解決方案。
+
+**學習體驗**
+
+最後想說的是，課程整體規劃真的很用心，小到 tag、作業與作業的關聯性跟順序安排，大到一個遊戲跟平台給大家練習，每一個細節都能感受到安排，是市面上我有用過的平台或課程們裡面最強的線上學習體驗。
```
Comment (aszx87410) [inline homeworks/week5/hw1.md:354]: ❤️

File: homeworks/week5/hw1.md [hunk 6]
```diff
@@ -1,3 +1,381 @@
...
+
+---
+
+### 概念筆記待辦
+
+- 伺服器運作原理 e.g. 瀏覽器渲染
```
Comment (aszx87410) [inline homeworks/week5/hw1.md:360]: 這兩個有點衝突XD 一個前端一個後端
瀏覽器渲染那個上一期有一支影片有提到，但我忘記這一期哪一週會有了

**Comment by aszx87410** [comment]
感謝心得分享～恭喜完成 HTTP 小遊戲，等之後進度進入到第五週我會把這遊戲也對外公開讓其他人也可以玩，有空的時候也可以挑戰看看之後的關卡XD

File: homeworks/week5/httpGame/lv10.js (hunk 1 above)
Comment (ishin4554) [reply homeworks/week5/httpGame/lv10.js:40]: 但用程式解滿好玩的我有一度還想說我發那麼多 request 會不會被擋下來，可以設流量限制當進階題 XDD

File: homeworks/week5/hw1.md (hunk 6 above)
Comment (ishin4554) [reply homeworks/week5/hw1.md:360]: 喔喔瞭解，我在修改一下拆開！

File: homeworks/week5/hw1.md (hunk 4 above)
Comment (ishin4554) [reply homeworks/week5/hw1.md:297]: [打不倒的空氣人](https://lidemy-oj.netlify.com/problems/10) 但這邊的案例可能是我想太多，應該不是老師的 test case 不完整，因為我有測沒有給數字的空陣列，當初有點不確定會不會為 0，真的送出通過了才發現應該都會有值。
